archunit/src/main/java/com/tngtech/archunit/core/domain/JavaType.java

```diff
@@ -82,7 +92,113 @@ public interface JavaType extends HasName {
+                return CONTINUE;
+            }
+
+            @Override
+            public Result visitParameterizedType(JavaParameterizedType type) {
+                result.add(type.toErasure());
+                return CONTINUE;
+            }
+        });
+        return result.build();
+    }
+
+    /**
+     * Traverses through the signature of this {@link JavaType}.<br>
+     * This method considers the type signature as a tree,
+     * where e.g. a {@link JavaClass} is a simple leaf,
+     * but a {@link JavaParameterizedType} has the type as root and then
+     * branches out into its actual type arguments, which in turn can have type arguments
+     * or upper/lower bounds in case of {@link JavaTypeVariable} or {@link JavaWildcardType}.<br>
+     * The following is a simple visualization of such a signature tree:
+     * <pre><code>
+     * List&lt;Map&lt;? extends Serializable, String[]&gt;&gt;
+     *                    |
+     *    Map&lt;? extends Serializable, String[]&gt;
+     *              /                   \
+     *  ? extends Serializable         String[]
+     *            |                      |
+     *      Serializable               String
+     * </code></pre>
+     * For every node visited the respective method of the provided {@code visitor}
+     * will be invoked. At every step it is possible to continue the traversal
+     * by returning {@link SignatureVisitor.Result#CONTINUE CONTINUE} or stop at that point by
+     * returning {@link SignatureVisitor.Result#STOP STOP}.<br><br>
+     * Note that the traversal will continue to traverse bounds of type variables,
+     * even if that type variable isn't declared in this signature itself.<br>
+     * E.g. take the following scenario
+     * <pre><code>
+     * class Example&lt;T extends String&gt; {
+     *     T[] field;
+     * }</code></pre>
+     * Traversing the {@link JavaField#getType() field type} of {@code field} will continue
+     * down to the upper bounds of the type variable {@code T} and thus end at the type {@code String}.<br><br>
+     * Also, note that the traversal will not continue down the type parameters of a raw type
+     * declared in a signature.<br>
+     * E.g. given the signature {@code class Example<T extends Map>} the traversal would stop at
+     * {@code Map} and not traverse down the type parameters {@code K} and {@code V} of {@code Map}.
+     *
+     * @param visitor A {@link SignatureVisitor} to invoke for every encountered {@link JavaType}
+     *                while traversing this signature.
+     */
+    @PublicAPI(usage = ACCESS)
+    void traverseSignature(SignatureVisitor visitor);
+
+    /**
+     * @see #traverseSignature(SignatureVisitor)
+     */
+    @PublicAPI(usage = INHERITANCE)
+    interface SignatureVisitor {
+        default Result visitClass(JavaClass type) {
+            return CONTINUE;
+        }
+
+        default Result visitArrayType(JavaClass type) {
+            return CONTINUE;
```

Comment:
I wonder whether the default implementation should be
```suggestion
            return visitClass(type);
```

Reply:
Hmm, but that would seem strange to me. Because the contract should be that `visitClass` is never called with an array type, right? (Otherwise, why even have `visitClass` and `visitArrayType` separately in the visitor? If you can't rely on `isArray()` to return false whenever `visitClass` is called?)
The traversal behavior will just traverse down to the component type by default, right? So implementing `CONTINUE` will finally call `visitClass` with the base component type, and that seems a reasonable behavior if I e.g. only implement `visitClass` but not `visitArrayType` 🤷

Reply:
> Because the contract should be that visitClass is never called with an array type, right?

Let's document the contract. 😉

I was checking whether we could make more use of the new `SignatureVisitor` internally, and found `JavaClassDependencies.dependenciesOfType(JavaType)`. `SignatureVisitor` won't be a nice solution there, but I realized that there might be more use cases where one previously had
```
        if (javaType instanceof JavaClass) {
            // do something with `(JavaClass) javaType`
        } else if (javaType instanceof JavaParameterizedType) {
            // do something with `(JavaParameterizedType) javaType`
        } else if (javaType instanceof JavaWildcardType) {
            // do something with `(JavaWildcardType) javaType`
        }
```
and it might not immediately obvious that refactoring this to the `SignatureVisitor` pattern requires overwriting both `visitClass` and `visitArrayType` for the ``do something with `(JavaClass) javaType`﻿`` part.

So yes, the question is actually whether `SignatureVisitor` should have individual methods for non-arrays and array `JavaClass`es. Having only one might be more consistent (why just distinguish arrays, and not, e.g. interfaces?) and therefore less confusing.

Reply:
Hmm, I see your point, I just figured it would be kind of nice that the default behavior would traverse up the component types and automatically only give you the element type (as I found the JLS calls our "base component type" 🤪). But maybe that in the end creates a more confusing API instead of helping 🤔

Reply:
One argument for the separate `visitArrayType` method might be that it's somewhat consistent with `visitGenericArrayType` 🤔 But I guess there the class hierarchy really is also distinct 🤷 
Guess I'm gonna remove the array type method then...